mpcs/src/basefold/encoding/rs.rs

```diff
@@ -272,90 +100,168 @@ where
+                }
+            })
+            .collect_vec();
+
         Ok((
             Self::ProverParameters {
-                fft_root_table: pp.fft_root_table,
-                gamma_powers: gamma_powers.clone(),
-                gamma_powers_inv_div_two: gamma_powers_inv.clone(),
+                dft: prover_dft,
+                t_inv_halves: t_inv_halves_prover.clone(),
                 full_message_size_log: max_message_size_log,
             },
             Self::VerifierParameters {
-                fft_root_table: verifier_fft_root_table,
+                dft: verifier_dft,
+                // TODO make verifier calculate fft root by itself
+                t_inv_halves: t_inv_halves_prover,
                 full_message_size_log: max_message_size_log,
-                gamma_powers,
-                gamma_powers_inv_div_two: gamma_powers_inv,
             },
         ))
     }
 
-    fn encode(pp: &Self::ProverParameters, coeffs: &FieldType<E>) -> FieldType<E> {
-        assert!(log2_strict(coeffs.len()) >= Spec::get_basecode_msg_size_log());
-        // Use the full message size to determine the shift factor.
-        Self::encode_internal(&pp.fft_root_table, coeffs, pp.full_message_size_log)
+    fn encode(pp: &Self::ProverParameters, rmm: RowMajorMatrix<E::BaseField>) -> Self::EncodedData {
+        // bh_evals is just a copy of poly.evals().
+        // Note that this function implicitly assumes that the size of poly.evals() is a
+        // power of two. Otherwise, the function crashes with index out of bound.
+        let num_vars = rmm.num_vars();
+        let num_polys = rmm.width();
+        if num_vars > pp.get_max_message_size_log() {
+            return PolyEvalsCodeword::TooBig(num_vars);
+        }
+
+        // In this case, the polynomial is so small that the opening is trivial.
+        // So we just build the Merkle tree over the polynomial evaluations.
+        // No codeword is needed.
+        if num_vars <= Spec::get_basecode_msg_size_log() {
+            return PolyEvalsCodeword::TooSmall(Box::new(rmm.into_default_padded_p3_rmm()));
+        }
+
+        // here 2 resize happend. first is padding to next pow2 height, second is extend to 2^get_rate_log times size
+        let mut m = rmm.into_default_padded_p3_rmm().to_row_major_matrix();
+        m.pad_to_height(m.height() * (1 << Spec::get_rate_log()), E::BaseField::ZERO);
+        let codeword = pp
+            .dft
+            .dft_batch(m)
+            // The encoding scheme always folds the codeword in left-and-right
+            // manner. However, in query phase the two folded positions are
+            // always opened together, so it will be more efficient if the
+            // folded positions are simultaneously sibling nodes in the Merkle
+            // tree. Therefore, instead of left-and-right folding, we bit-reverse
+            // the codeword to make the folding even-and-odd, i.e., adjacent
+            // positions are folded.
```

Comment:
I think the reason for using bit reversed form of original multilinear polynomial is to have better locality in the commit/fold phase not to optimize the query phase as the number of rows to query is less than 200.

Reply:
Using bit reversed form is beneficial for commit phase and query phase in the `pcs::open()`. For commit phase, we can fold the long input by half using `par_chunks_mut(2)`.

Reply:
Yeah added in commit https://github.com/scroll-tech/ceno/pull/843/commits/17f9787748578b2169c93ca274cf12526a0ea7ec